rekall-core/rekall/plugins/linux/heap_analysis.py

```diff
@@ -0,0 +1,4388 @@
+
+        #############################################
+
+    # Goes to the top chunk of a given arena, gets its heap_info offset and follows all
+    # _heap_info.prev members until the last one (for the last _heap_info, the prev field is 0x0
+    def _heaps_for_arena(self, arena):
+        """Returns a sorted list of all heap_info structs for a given arena:
+        [0] = first heap_info.
+        This method is normally only called on initialization for a new task and
+        further access to heaps is done via the heaps attribute of each arena.
+        """
+
+        heap_infos = list()
+
+        if arena.top_chunk:
+            last_heap_info = self.__heap_for_ptr(arena.top_chunk)
+
+            if not last_heap_info.ar_ptr.dereference() == arena:
+                self.session.logging.error("Unexpected error: current heap_info's arena pointer "
+                                           "doesn't point to the expected arena. Maybe wrong "
+                                           "profile or different cause.")
+
+            heap_infos = list(iterate_through_linked_list(last_heap_info,
+                                                          lambda x: x.prev))[::-1]
+
+        return heap_infos
+
+
+    def get_main_arena(self):
+        """Returns the main_arena for the current task, which is the first arena in the
+        arenas list. If the current instance is not intialized, it logs a warning.
+        """
+
+        if self.arenas:
+            if self.arenas[0].is_main_arena:
+                return self.arenas[0]
+
+            else:
+                self.session.logging.warn(
+                    "First arena in the arenas list doesn't seem to be the main_arena.")
+
+        else:
+            self.session.logging.warn(
+                "There are no arenas. Maybe this instance has not been initialized"
+                " for the current task. Try to initialize it via 'init_for_task'.")
+
+        return None
+
+
+    def _initialize_arenas(self, main_arena):
+        """Gathers all arenas, their heaps and sets main_arenas first chunk."""
+
+        self._initialize_main_arena(main_arena)
+
+        for arena in iterate_through_linked_list(main_arena, lambda x: x.next):
+            self.arenas.append(arena)
+            arena.freed_fast_chunks = list(arena.get_freed_chunks_fastbins())
+            arena.freed_chunks = list(arena.get_freed_chunks_bins())
+
+            if not arena.is_main_arena:
+                temp_top = arena.top.dereference()
+                temp_top.is_top_chunk = True
+                arena.top_chunk = temp_top
+                arena.heaps = self._heaps_for_arena(arena)
+                # in this implementation, thread arenas don't use the first_chunk member,
+                # but their heaps do
+                arena.first_chunk = None
+                arena.is_main_arena = False
+                arena.mmapped_first_chunks = None
+
+
+    def _initialize_main_arena(self, main_arena):
```

Comment:
there is only an arena class (malloc_state) and the main_arena is an instance of it, but also thread arenas are
i will have a look if i can merge both initializer methods within this class